src/libsystemd/sd-device/sd-device.c

```diff
@@ -245,69 +245,65 @@ _public_ int sd_device_new_from_devnum(sd_device **ret, char type, dev_t devnum)
 }
 
 _public_ int sd_device_new_from_subsystem_sysname(sd_device **ret, const char *subsystem, const char *sysname) {
-        char syspath[PATH_MAX], *name;
+        char *name, *syspath;
 
         assert_return(ret, -EINVAL);
-        assert_return(subsystem, -EINVAL);
-        assert_return(sysname, -EINVAL);
-        assert_return(strlen(sysname) < PATH_MAX - strlen("/sys/bus/"), -ENAMETOOLONG);
+        assert_return(filename_is_valid(subsystem), -EINVAL);
+        assert_return(filename_is_valid(sysname), -EINVAL);
 
         if (streq(subsystem, "subsystem")) {
-                if (snprintf_ok(syspath, sizeof syspath, "/sys/subsystem/%s", sysname) &&
-                    access(syspath, F_OK) >= 0)
+                syspath = strjoina("/sys/subsystem/", sysname);
+                if (access(syspath, F_OK) >= 0)
                         return sd_device_new_from_syspath(ret, syspath);
 
-                if (snprintf_ok(syspath, sizeof syspath, "/sys/bus/%s", sysname) &&
-                    access(syspath, F_OK) >= 0)
+                syspath = strjoina("/sys/bus/", sysname);
+                if (access(syspath, F_OK) >= 0)
                         return sd_device_new_from_syspath(ret, syspath);
```

Comment:
The problem with this approach is that the repeated `strjoina`s "add up", so effectively we're pushing the string 10+ times onto the stack. And the fixed PATH_MAX allocation actually is not a problem: the allocation is just one machine instruction and doesn't matter.

Potentially, a important difference is in calling `snprinf` vs. `strjoina`. But it seems to be less than one would expect:
```diff
diff --git src/test/test-util.c src/test/test-util.c
index 0fe7a381c7..1a098a88e9 100644
--- src/test/test-util.c
+++ src/test/test-util.c
@@ -502,6 +502,38 @@ static void test_ptr_to_int(void) {
         assert_se(PTR_TO_INT(INT_TO_PTR(INT_MIN)) == INT_MIN);
 }
 
+static size_t do_it(const char *arg) {
+        const char *s;
+
+        s = strjoina(arg, "/", arg, "/", arg);
+        return strlen(s);
+}
+
+static void test_strjoina(const char *arg) {
+        usec_t then = now(CLOCK_MONOTONIC);
+
+        for (unsigned i = 0; i < 10000000; i++)
+                do_it(arg);
+
+        log_info("%s: %zu", __func__, now(CLOCK_MONOTONIC) - then);
+}
+
+static size_t do_it2(const char *arg) {
+        char s[PATH_MAX];
+
+        snprintf(s, sizeof s, "%s/%s/%s", arg, arg, arg);
+        return strlen(s);
+}
+
+static void test_snprintf(const char *arg) {
+        usec_t then = now(CLOCK_MONOTONIC);
+
+        for (unsigned i = 0; i < 10000000; i++)
+                do_it2(arg);
+
+        log_info("%s: %zu", __func__, now(CLOCK_MONOTONIC) - then);
+}
+
 int main(int argc, char *argv[]) {
         test_setup_logging(LOG_INFO);
 
@@ -523,5 +555,8 @@ int main(int argc, char *argv[]) {
         test_foreach_pointer();
         test_ptr_to_int();
 
+        test_strjoina(argv[0]);
+        test_snprintf(argv[0]);
+
         return 0;
 }
```

I get 0.7 vs 1.4 fairly consistently, which means that snprintf is only two times slower. Considering that this is not a hot path and the runtime of this function is going to be dominated by the syscalls anyway, I think 2× overhead for this is acceptable.